moodler/upload.py

```diff
@@ -0,0 +1,252 @@
+
+logger = logging.getLogger(__name__)
+
+GRADING_WORKSHEET_UPLOAD_CONTEXTID_PATTERN = (
+    r'<input name="contextid" type="hidden" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_SESSKEY_PATTERN = (
+    r'<input name="sesskey" type="hidden" ' r'value="([\w\d]+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_GRADESFILE_PATTERN = (
+    r'<input type="hidden" name="gradesfile" id="id_gradesfile" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_IMPORTID_PATTERN = (
+    r'<input name="importid" type="hidden" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_DRAFTID_PATTERN = (
+    r'<input name="draftid" type="hidden" ' r'value="(\d+)"'
+)
+
+SUCCESS_GRADING_WORKSHEET_UPLOAD_PATTERN = r"Updated (\d+) grades and feedback"
+
+
+class UploadException(MoodlerException):
+    pass
+
+
+class InvalidGradingPage(UploadException):
+    pass
+
+
+class InvalidUploadGradingWorksheetPage(UploadException):
+    pass
+
+
+class InvalidUploadGradingConfirmationPage(UploadException):
+    pass
+
+
+def get_params_from_grading_page(
+    assignment_id: str, session: requests.Session
+) -> Tuple[str, str]:
+    # Build the get request to the grading page.
+    params = {"id": assignment_id, "action": "grading"}
+    response = session.get(URL + "/mod/assign/view.php", params=params)
+
+    grading_page_content = response.content.decode()
+    contextid_match = re.search(
+        GRADING_WORKSHEET_UPLOAD_CONTEXTID_PATTERN, grading_page_content
+    )
+    sesskey_match = re.search(
+        GRADING_WORKSHEET_UPLOAD_SESSKEY_PATTERN, grading_page_content
+    )
+
+    if contextid_match is None:
+        raise InvalidGradingPage(
+            "The contextid required to upload the "
+            "grading sheet to moodle was not found."
+        )
+    if sesskey_match is None:
+        raise InvalidGradingPage(
+            "The sesskey required to upload the "
+            "grading sheet to moodle was not found."
+        )
+    return contextid_match.group(1), sesskey_match.group(1)
+
+
+def get_params_from_upload_grading_worksheet_page(
+    assignment_id: str, session: requests.Session
+) -> str:
+    # Build the get request to the upload grading worksheet page.
+    params = {
+        "id": assignment_id,
+        "plugin": "offline",
+        "pluginsubtype": "assignfeedback",
+        "action": "viewpluginpage",
+        "pluginaction": "uploadgrades",
+    }
+    response = session.get(URL + "/mod/assign/view.php", params=params)
+
+    upload_grading_worksheet_page_content = response.content.decode()
+    gradesfile_match = re.search(
+        GRADING_WORKSHEET_UPLOAD_GRADESFILE_PATTERN,
+        upload_grading_worksheet_page_content,
+    )
+
+    if gradesfile_match is None:
+        raise InvalidUploadGradingWorksheetPage(
+            "The gradesfile required to upload the "
+            "grading sheet to moodle was not found."
+        )
+    return gradesfile_match.group(1)
+
+
+def upload_file_to_moodle(
+    file_path: str,
+    content_type: str,
```

Comment:
let's make the content type hardcoded for now, which I prefer until we need it to be generic

**OR**
move it to an enum to avoid issues

```
from enum import Enum
class ContentType(Enum):
  Excel = "application/vnd.ms-excel"

content_type = ContentType.Excel
```

And then use `content_type.value`

Reply:
Like!